### PR TITLE
Final touches for stable runtime :see_no_evil: 

### PR DESCRIPTION
**Pull request by lluiscampos**
Both changes are provisional and we have tasks to follow-up. See individual commits :popcorn:

### CODE REVIEW

File: core/src/mender-client.c [hunk 1]
```diff
@@ -1165,17 +1009,30 @@ mender_client_update_work_function(void) {
...
+    mender_log_info("Getting device type");
+
     /* Retrieve device type from artifact */
     const char *device_type_artifact = NULL;
     if (MENDER_OK != (ret = mender_artifact_get_device_type(mender_artifact_ctx, &device_type_artifact))) {
         mender_log_error("Unable to get device type from artifact");
-        mender_client_publish_deployment_status(deployment->id, MENDER_DEPLOYMENT_STATUS_FAILURE);
-        if (mender_client_deployment_needs_set_pending_image) {
-            mender_flash_abort_deployment(mender_client_flash_handle);
+        if (true == only_for_testing) {
```
Comment (larsewi) [inline core/src/mender-client.c:1020]: No need to compare boolean value with `true`

File: core/src/mender-client.c [hunk 2]
```diff
@@ -1165,17 +1009,30 @@ mender_client_update_work_function(void) {
...
+            mender_log_error("Will be using PROJECT_NAME as device type for testing purposes");
+            char *device_type_test = malloc(strlen("PROJECT_NAME") + 1);
+            strcpy(device_type_test, "PROJECT_NAME");
```
Comment (larsewi) [inline core/src/mender-client.c:1023]:
```suggestion
            char *device_type_test = strdup("PROJECT_NAME");
```

File: core/src/mender-client.c [hunk 3]
```diff
@@ -1272,10 +1132,13 @@ mender_client_update_work_function(void) {
...
         mender_client_deployment_data = NULL;
     }
-    mender_artifact_release_ctx(mender_artifact_ctx);
+    mender_log_info("Releasing artifact context");
+    // we will be restarting anyways
+    // mender_artifact_release_ctx(mender_artifact_ctx);
 
     /* Check if the system must restart following downloading the deployment */
     if (true == mender_client_deployment_needs_restart) {
+        mender_log_info("All righty; get it done baby!");
```
Comment (larsewi) [inline core/src/mender-client.c:1141]: 😆

**Review by larsewi — APPROVED** [review]
🚀

File: core/src/mender-client.c [hunk 4]
```diff
@@ -655,65 +519,51 @@ mender_client_exit(void) {
...
+            }
+            mender_client_state = MENDER_CLIENT_STATE_AUTHENTICATED;
+            break;
+        case MENDER_CLIENT_STATE_AUTHENTICATED:
+            mender_log_info("work function; authenticated");
+            ret = mender_client_update_work_function();
+            break;
+        default:
+            mender_log_info("work function; default");
+            break;
```
Comment (vpodzime) [inline core/src/mender-client.c:563]: Please avoid using `default`, we should explicitly cover all possible values of the enum.

File: core/src/mender-client.c [hunk 5]
```diff
@@ -655,65 +519,51 @@ mender_client_exit(void) {
...
-    mender_client_network_release();
+    switch (mender_client_state) {
+        case MENDER_CLIENT_STATE_INITIALIZATION:
+            mender_log_info("work function; initialization");
+            if (MENDER_DONE != (ret = mender_client_initialization_work_function())) {
+                return ret;
+            }
+            /* Update client state */
+            mender_client_state = MENDER_CLIENT_STATE_AUTHENTICATION;
+            break;
```
Comment (vpodzime) [inline core/src/mender-client.c:548]: why `break` here? Can't we just continue without waiting for the next execution of this function hitting this `switch()` again?

File: core/src/mender-client.c (hunk 2 above)
Comment (vpodzime) [reply core/src/mender-client.c:1023]: Plus check that it was actually allocated as is not `NULL`.

File: core/src/mender-client.c (hunk 3 above)
Comment (vpodzime) [reply core/src/mender-client.c:1141]: :joy:

**Review by vpodzime — COMMENTED** [review]
I wish I could say it looks *good* to me, but it's a great step for the Mender-kind! Hence :+1: :beers:

File: core/src/mender-client.c (hunk 1 above)
Comment (lluiscampos) [reply core/src/mender-client.c:1020]: Actually, I have to remove this. Thanks

File: core/src/mender-client.c (hunk 2 above)
Comment (lluiscampos) [reply core/src/mender-client.c:1023]: Thanks

File: core/src/mender-client.c (hunk 3 above)
Comment (lluiscampos) [reply core/src/mender-client.c:1141]: @pasinskim words...

File: core/src/mender-client.c (hunk 5 above)
Comment (lluiscampos) [reply core/src/mender-client.c:548]: Yes, I'll do that.

File: core/src/mender-client.c (hunk 1 above)
Comment (lluiscampos) [reply core/src/mender-client.c:1020]: (I mean, you can see that it is just set above to `true` and never modified...`

File: core/src/mender-client.c (hunk 2 above)
Comment (lluiscampos) [reply core/src/mender-client.c:1023]: This code is removed. See above comment.

File: core/src/mender-client.c (hunk 4 above)
Comment (lluiscampos) [reply core/src/mender-client.c:563]: We are already going through all of them. Do you mean completely removing the `default` or adding an assert to make sure we never have an untreated case?

(this is a question in general, as I know you've been commenting on these before)

Comment (lluiscampos) [reply core/src/mender-client.c]: > (...) completely removing the `default` (...)

I did this

File: core/src/mender-client.c [hunk 6]
```diff
@@ -1272,10 +1117,13 @@ mender_client_update_work_function(void) {
         cJSON_Delete(mender_client_deployment_data);
         mender_client_deployment_data = NULL;
     }
-    mender_artifact_release_ctx(mender_artifact_ctx);
+    mender_log_info("Releasing artifact context");
+    // we will be restarting anyways
```
Comment (pasinskim) [inline core/src/mender-client.c:1121]: Maybe you can consider adding a `TODO` comment that we should check the context releasing (deinitialization; cleanup) before the restart.

**Comment by pasinskim** [comment]
Nice. Thanks for tidying it up!

File: core/src/mender-client.c [hunk 7]
```diff
@@ -376,12 +340,6 @@ mender_client_register_artifact_type(char *type,
     mender_client_artifact_type_t **tmp;
     mender_err_t                    ret;
```
Comment (lluiscampos) [inline core/src/mender-client.c:341]: Note to self,
```suggestion
    mender_err_t                    ret = MENDER_OK;
```

Here we are missing initialization. Please **review all of these**

File: core/src/mender-client.c (hunk 4 above)
Comment (vpodzime) [reply core/src/mender-client.c]: Yes, completely removing, the compiler will then warn us if there is an unhandled case.

**Review by vpodzime — APPROVED** [review]

File: platform/scheduler/posix/src/mender-scheduler.c [hunk 8]
```diff
@@ -458,3 +458,22 @@ mender_scheduler_work_queue_thread(void *arg) {
...
     pthread_exit(NULL);
 }
+
+/**
+ * @brief Dummy implementation
+ */
+mender_err_t
+mender_scheduler_alt_work_create(mender_scheduler_alt_work_function_t func, int32_t interval) {
+    (void)func;
+    (void)interval;
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply platform/scheduler/posix/src/mender-scheduler.c]: [Not really](https://gitlab.com/Northern.tech/Mender/mender-mcu/-/jobs/7813220664) :sweat_smile:

**Pull request merged**